src/Microsoft.TemplateEngine.Edge/Settings/SettingsLoader.cs

```diff
@@ -224,77 +279,104 @@ public void Dispose()
                     }
                 }
             }
+        }
 
-            JObject parsed;
-            using (Timing.Over(_environmentSettings.Host, "Parse settings"))
+        /// <summary>
+        /// Loads the settings, should be called under locking <see cref="_settingsLock"/>.
+        /// </summary>
+        private void EnsureLoaded()
+        {
+            lock (_settingsLock)
             {
-                try
+                if (_loaded)
                 {
-                    parsed = JObject.Parse(userSettings);
+                    return;
                 }
-                catch (Exception ex)
+
+                string? userSettings = null;
+                using (Timing.Over(_environmentSettings.Host, "Read settings"))
                 {
-                    throw new EngineInitializationException("Error parsing the user settings file", "Settings File", ex);
+                    for (int i = 0; i < MaxLoadAttempts; ++i)
+                    {
+                        try
+                        {
+                            userSettings = _paths.ReadAllText(_paths.SettingsFile, "{}");
+                            break;
+                        }
+                        catch (IOException)
+                        {
+                            if (i == MaxLoadAttempts - 1)
+                            {
+                                throw;
+                            }
+
+                            Task.Delay(2).Wait();
+                        }
+                    }
                 }
-            }
 
-            using (Timing.Over(_environmentSettings.Host, "Deserialize user settings"))
-            {
-                _userSettings = new SettingsStore(parsed);
-            }
+                JObject parsed;
+                using (Timing.Over(_environmentSettings.Host, "Parse settings"))
+                {
+                    try
+                    {
+                        parsed = JObject.Parse(userSettings);
+                    }
+                    catch (Exception ex)
+                    {
+                        throw new EngineInitializationException("Error parsing the user settings file", "Settings File", ex);
+                    }
+                }
 
-            using (Timing.Over(_environmentSettings.Host, "Init probing paths"))
-            {
-                if (_userSettings.ProbingPaths.Count == 0)
+                using (Timing.Over(_environmentSettings.Host, "Deserialize user settings"))
                 {
-                    _userSettings.ProbingPaths.Add(_paths.User.Content);
+                    _userSettings = new SettingsStore(parsed);
                 }
-            }
 
-            using (Timing.Over(_environmentSettings.Host, "Init Component manager"))
-            {
-                _componentManager = new ComponentManager(this, _userSettings);
-            }
+                using (Timing.Over(_environmentSettings.Host, "Init probing paths"))
```

Comment:
Again, if we are only adding some object to a list, measured time would be irrelevant.